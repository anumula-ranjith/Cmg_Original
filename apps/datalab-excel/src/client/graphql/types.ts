export type IQueryPath = Readonly<{
  [key: string]: string | IQueryPath;
}>;

export type GQLScalar = string | number | boolean | null;

export type GQLEnum = Record<string, GQLScalar>;

export type GQLNode = Readonly<{
  [key: string]: GQLScalar | GQLScalar[] | GQLNode | GQLNode[] | GQLEnum;
}>;

export type GQLPageInfo = Readonly<{
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}>;

export type GQLQueryType = Readonly<{
  pageInfo?: GQLPageInfo;
  totalCount?: number;
  items: GQLNode[];
}>;

export type GQLResponse = Readonly<{
  data?: Record<string, GQLQueryType | GQLSchema | GQLSchemaForTypeQuery> | GQLSchemaKind;
  errors?: GQLError[];
}>;

export type GQLLocation = Readonly<{
  line: number;
  column: number;
}>;

export type GQLErrorExtensions = Readonly<{
  type: string;
  field: string;
  responseName: string;
  specifiedBy: string;
}>;

export type GQLError = Readonly<{
  message: string;
  locations: GQLLocation[];
  extensions: GQLErrorExtensions;
}>;

export type GQLTypeRef = Readonly<{
  kind: string;
  name?: string;
  ofType?: GQLTypeRef;
  fields?: GQLField[];
}>;

export type GQLField = Readonly<{
  name: string;
  type: GQLTypeRef;
  description: string;
}>;

export type GQLType = Readonly<{
  kind: string;
  name: string;
  description?: string;
  fields?: GQLField[];
  interfaces?: GQLTypeRef[];
  possibleTypes?: GQLTypeRef[];
  enumValues?: { name: string; description?: string }[];
  inputFields?: GQLField[];
}>;

export type GQLSchema = Readonly<{
  queryType: GQLTypeRef;
  mutationType?: GQLTypeRef;
  subscriptionType?: GQLTypeRef;
  types: GQLType[];
  directives: {
    name: string;
    description?: string;
    locations: string[];
    args: GQLField[];
  }[];
}>;

export type GQLSchemaForTypeQuery = Readonly<{
  __type: {
    name: string;
    fields?: GQLField[];
  };
}>;

export type GQLSchemaKind = Readonly<{
  __schema: {
    name: string;
    kind: string;
    queryType: { fields: GQLField[] };
  };
}>;
