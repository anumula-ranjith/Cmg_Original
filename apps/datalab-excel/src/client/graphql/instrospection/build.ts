import {
  buildOperationTree,
  findOperationRootTypeOrThrow,
} from '@cmg/graphql-toolkit-introspection';
import type { GraphQLSchema } from 'graphql';

import { toCamelCase } from '../../../utilities/stringutils';
import type { CoerceFieldsIntoOperationOptions } from './types';

export type Query = {
  take: number;
  skip: number;
};

export type BuildQueryOptions = {
  query: Query;
  build?: CoerceFieldsIntoOperationOptions;
};

const getQueryName = (resourceType: string) => toCamelCase(resourceType);

export const buildQuery = (
  schema: GraphQLSchema,
  source: string,
  options: BuildQueryOptions,
  whereClause: string | null | undefined
) => {
  const rootType = findOperationRootTypeOrThrow(schema, 'query', getQueryName(source));
  const document = buildOperationTree(rootType, options.build);
  const take = `take: ${options.query.take},`;
  const skip = `skip: ${options.query.skip},`;
  const where = whereClause ? `where: ${whereClause},` : '';
  const query = `{${source}(${take}${skip}${where}) ${document}}`;

  return query;
};
