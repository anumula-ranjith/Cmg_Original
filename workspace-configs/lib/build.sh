#!/usr/bin/env sh

#
# Library package build
#
# This script generates typescript declarations and a set of es modules + cjs modules of a particular
# workspace's source.
# Usage: Call this from a workspace yarn script (e.g. packages/ecm-client-web/package.json)
# Files:
#  * tsconfig.build.json - file path is relative to location the script is called from, which should be the workspace root.
#  * .babelrc.js/json - not referenced here but implicitly used by babel commands. this file also sits in a workspace root.
#
tsc --project tsconfig.build.json \
  && BABEL_ENV=production BABEL_MODULE_MODE=es babel src -d --out-dir dist/es --copy-files --no-copy-ignored --extensions '.js,.ts,.tsx,.snap' \
  && BABEL_ENV=production BABEL_MODULE_MODE=cjs babel src -d --out-dir dist/lib --copy-files --no-copy-ignored --extensions '.js,.ts,.tsx,.snap'
