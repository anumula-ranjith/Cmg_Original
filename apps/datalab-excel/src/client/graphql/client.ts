import type { GraphQLSchema } from 'graphql';

import type { BuildQueryOptions } from './instrospection/build';
import { buildQuery } from './instrospection/build';
import type { GQLError, GQLQueryType, GQLResponse } from './types';

export async function fetchInstrospection(
  graphQlUrl: string,
  apiKey: string | null,
  query: string,
  variables: Record<string, string> | null = null
): Promise<GQLResponse> {
  if (!apiKey) {
    throw new Error('No API Key set');
  }
  const response = await fetch(graphQlUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      query: query,
      variables: variables,
    }),
  });
  if (response.status !== 200) {
    throw new Error(`Server responded with ${response.status} ${response.statusText}`);
  }

  const contentType = response.headers.get('content-type');
  if (!contentType || !contentType.includes('application/graphql-response+json')) {
    throw new Error(
      `Expected JSON response but received ${contentType ? contentType : 'no content type'}`
    );
  }
  const result = (await response.json()) as GQLResponse;
  if (result.errors) {
    const messages = result.errors.map((error: GQLError) => error.message).join('\n');
    throw new Error(`Errors: ${messages}`);
  }
  return result;
}

export async function fetchGraphqlData(
  schema: GraphQLSchema,
  graphQlUrl: string,
  apiKey: string | null,
  source: string,
  queryOptions: BuildQueryOptions,
  whereClause: string | null | undefined,
): Promise<GQLQueryType> {
  if (!apiKey) {
    throw new Error('No API key found');
  }
  const query = buildQuery(schema, source, queryOptions, whereClause);
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      query,
    }),
  };
  const response = await fetch(graphQlUrl, options);

  if (response.status !== 200) {
    throw new Error(`Server responded with ${response.status} ${response.statusText}`);
  }
  const result = (await response.json()) as GQLResponse;
  if (result.errors) {
    const messages = result.errors.map(error => error.message).join('<br />');
    throw new Error(`GraphQL: ${messages}`);
  }
  const typeData = (result.data ? result.data[source] : { items: [] }) as GQLQueryType;
  return typeData;
}
