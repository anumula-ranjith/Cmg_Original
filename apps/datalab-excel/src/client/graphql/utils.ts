import { findOperationRootTypeOrThrow } from '@cmg/graphql-toolkit-introspection';
import type { GraphQLOutputType, GraphQLSchema } from 'graphql';
import { GraphQLEnumType, getNamedType, isObjectType, isScalarType } from 'graphql';

import type { TableCell } from '../../excel/types';
import type { CoerceFieldsIntoOperationOptions } from './instrospection/types';
import type { GQLNode } from './types';

export const DEFAULT_MAX_NESTED_COERCION = 2;

export function escapeIfNeeded(s: string | object): string {
  let str: string;
  if (typeof s === 'object') {
    str = JSON.stringify(s).replace(/[{}"]/g, m => {
      switch (m) {
        case '{':
          return '\\{';
        case '}':
          return '\\}';
        case '"':
          return '\\"';
        default:
          return m;
      }
    });
  } else {
    str = s.toString();
  }
  return `"${str}"`;
}

export function stringifyResponse(response: unknown): TableCell {
  if (response === null || response === undefined) {
    return '';
  }
  if (typeof response === 'boolean') {
    return response ? 'TRUE' : 'FALSE';
  }
  if (typeof response === 'number') {
    return response;
  }
  if (typeof response === 'string') {
    return response.length < 1 ? '' : response;
  }
  if (typeof response === 'object') {
    if (Array.isArray(response)) {
      if (response.length === 0) {
        return '';
      }
    } else {
      return escapeIfNeeded(JSON.stringify(response));
    }
    return JSON.stringify(response);
  }
  return response as string;
}

export function parseGraphQlResponse(
  response: GQLNode[],
  userColumns: Record<string, string>,
  headers: string[]
): Record<string, TableCell>[] {
  const userColumnsKeys = Object.keys(userColumns);
  const result = response.map((item: GQLNode) => {
    const rows: Record<string, TableCell> = {};
    headers.forEach((header: string) => {
      if (userColumnsKeys.includes(header)) {
        rows[header] = userColumns[header];
      } else {
        let value: unknown = item;
        const headerParts = header.split('.');
        for (const key of headerParts) {
          if (Array.isArray(value)) {
            value = JSON.stringify(
              value.map(item => {
                const v = item as Record<string, unknown>;
                return `${(v[key] ? v[key] : '') as string}`;
              })
            );
          } else if (value !== null && typeof value === 'object') {
            value = value[key as keyof typeof value];
          }
        }
        const answer = stringifyResponse(value);
        rows[header] = answer;
      }
    });
    return rows;
  });
  return result;
}

const MAX_DEPTH = 3;

export type SourceField = {
  [key: string]: SourceField | null;
};

const retrieveSourceFieldsFromSchema = (
  type: GraphQLOutputType,
  options?: CoerceFieldsIntoOperationOptions,
  path = '',
  depth = 0
): SourceField => {
  if (depth >= (options?.maxNested ?? MAX_DEPTH)) {
    return {};
  }
  const result: SourceField = {};
  const namedType = getNamedType(type);

  if (isObjectType(namedType)) {
    const fields = namedType.getFields();
    for (const fieldName in fields) {
      const field = fields[fieldName];
      const currPath = path ? `${path}.${fieldName}` : fieldName;
      const fieldNamedType = getNamedType(field.type);
      const isLeaf = isScalarType(fieldNamedType);
      const isObject = isObjectType(fieldNamedType);

      const isIncluded =
        options?.include == null ||
        options.include.some(includePath =>
          isLeaf
            ? currPath.startsWith(`${includePath}.`) || currPath === includePath
            : includePath.startsWith(currPath)
        );

      const isExcluded = options?.exclude ? options.exclude.includes(currPath) : false;
      if (isIncluded && !isExcluded) {
        if (isObject) {
          const nestedObject = retrieveSourceFieldsFromSchema(
            field.type,
            options,
            currPath,
            depth + 1
          );
          if (Object.keys(nestedObject).length > 0) {
            result[fieldName] = nestedObject;
          }
        } else {
          result[fieldName] = null;
        }
      }
    }
  }

  return result;
};

export const getSourceFields = (
  schema: GraphQLSchema,
  source: string,
  options?: CoerceFieldsIntoOperationOptions
): SourceField => {
  const rootType = findOperationRootTypeOrThrow(schema, 'query', source);
  const namedType = getNamedType(rootType);
  const fields = isObjectType(namedType) ? Object.entries(namedType.getFields()) : [];
  const items = fields.filter(([fieldName]) => {
    return fieldName === 'items';
  });
  if (items.length > 0) {
    const itemsType = items[0][1].type;
    return retrieveSourceFieldsFromSchema(itemsType, options);
  }
  return {};
};