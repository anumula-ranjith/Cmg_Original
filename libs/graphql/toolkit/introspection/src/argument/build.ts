import { filterNotNull } from '@cmg/core-array';
import { findArgumentsByField } from '@cmg/graphql-toolkit-core';
import type { ArgumentNode } from 'graphql';
import { astFromValue, isInputType } from 'graphql';

import type { BuildArgumentNodeParams, CoerceArgumentNodeByFieldParams } from './types';

/**
 * Builds an GraphQL AST Argument Node from a GraphQL Type.
 * @param params Build Parameters.
 * @returns GraphQL AST Argument Node
 */
export const buildArgumentNode = (params: BuildArgumentNodeParams): ArgumentNode | undefined => {
  const { schema, name, value, type } = params;
  if (value === undefined) {
    return undefined;
  }

  const inputType = typeof type === 'string' ? schema.getType(type) : type;
  if (!isInputType(inputType)) {
    return undefined;
  }

  const astValue = astFromValue(value, inputType);
  if (astValue == null) {
    return undefined;
  }

  return {
    kind: 'Argument',
    name: { kind: 'Name', value: name },
    value: astValue,
  };
};

/**
 * Builds a collection of GraphQL AST Argument Node from a GraphQL Type's field.
 * @param params Build Parameters.
 * @returns A collection of GraphQL AST Argument Node.
 */
export const buildArgumentNodesByField = (
  params: CoerceArgumentNodeByFieldParams
): readonly ArgumentNode[] => {
  const { schema, field, typeName, args } = params;
  return findArgumentsByField(schema, typeName, field)
    .filter(arg => args?.[arg.name] != null)
    .map(arg =>
      buildArgumentNode({
        schema,
        name: arg.name,
        type: arg.type,
        value: args?.[arg.name],
      })
    )
    .filter(filterNotNull);
};
