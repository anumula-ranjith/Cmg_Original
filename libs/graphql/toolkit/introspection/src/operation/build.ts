import type {
  DocumentNode,
  GraphQLNamedType,
  GraphQLOutputType,
  GraphQLSchema,
  OperationDefinitionNode,
  OperationTypeNode,
  SelectionNode,
  SelectionSetNode,
} from 'graphql';
import { getNamedType, isObjectType, isScalarType, parse } from 'graphql';

import { buildArgumentNodesByField } from '../argument';
import { DEFAULT_MAX_NESTED_COERCION } from './const';
import { findOperationRootTypeOrThrow } from './find';
import type { BuildOperation, CoerceFieldsIntoOperationOptions } from './types';

/**
 * Builds a GraphQL Operation Structured AST Selection Tree.
 * @param type The GraphQL Type.
 * @param options Options.
 * @returns Returns a Structured AST Selection Tree Definition as a string.
 */
export const buildOperationTree = (
  type: GraphQLOutputType,
  options?: CoerceFieldsIntoOperationOptions,
  path?: string,
  depth = 0,
  _cache?: Map<GraphQLNamedType, number>
): string => {
  const map = _cache ?? new Map<GraphQLNamedType, number>();
  const namedType = getNamedType(type);
  const maxNested = options?.maxNested ?? DEFAULT_MAX_NESTED_COERCION;
  const nestedCount = map.get(namedType) ?? 0;

  if (options?.maxCyclicNested && depth > options.maxCyclicNested) {
    return '';
  }

  if (isObjectType(namedType)) {
    map.set(namedType, nestedCount + 1);
    const fieldEntries = Object.entries(namedType.getFields());
    const fields = fieldEntries
      .filter(([, fieldType]) => {
        const fieldNamedType = getNamedType(fieldType.type);
        const fieldNestedCount = map.get(fieldNamedType) ?? 0;
        return fieldNestedCount < maxNested;
      })
      .map(([, fieldType]) => {
        const fieldName = fieldType.name;
        const fieldEntries = Object.entries(namedType.getFields());
        const isObject = isObjectType(getNamedType(fieldType.type));
        const currPath = path == null ? fieldName : `${path}.${fieldName}`;
        const descendants = buildOperationTree(fieldType.type, options, currPath, depth + 1, map);
        if (isObject && descendants === '' && fieldEntries.length > 0) {
          return { coerced: null, currPath, fieldType };
        }
        return { coerced: `${fieldName}${descendants}`, currPath, fieldType };
      })
      .filter(({ coerced, currPath, fieldType }) => {
        const isLeafType = isScalarType(getNamedType(fieldType.type));
        const isIncluded =
          options?.include == null ||
          options.include.some(includePath =>
            isLeafType
              ? currPath.startsWith(`${includePath}.`) || currPath === includePath
              : includePath.startsWith(currPath)
          );
        const isExcluded = options?.exclude?.includes(currPath);
        return coerced !== null && isIncluded && !isExcluded;
      })
      .map(({ coerced }) => coerced);

    if (depth === 0) {
      fields.push('totalCount');
    }
    if (fields.length === 0) {
      return '';
    }

    return `{\n\t${fields.join(',\n\t')}\t\n}`;
  }

  return '';
};

/**
 * Builds a GraphQL Operation Selection Node from a GraphQL Type.
 * @param schema The GraphQL Schema.
 * @param operationType The Operation Type.
 * @param operation The Operation to build.
 * @returns An AST Selection Node.
 */
export const buildOperationSelectionNode = (
  schema: GraphQLSchema,
  operationType: OperationTypeNode,
  operation: BuildOperation
): SelectionNode => {
  const field = typeof operation === 'object' ? operation.field : operation;
  const options = typeof operation === 'object' ? operation.options : undefined;
  const alias = typeof operation === 'object' ? operation.alias : undefined;
  const operationTypeName =
    (operationType === 'query' ? schema.getQueryType()?.name : schema.getMutationType()?.name) ??
    'Query';

  const operationArgs = buildArgumentNodesByField({
    schema,
    typeName: operationTypeName,
    field,
    args: options?.args,
  });

  const rootType = findOperationRootTypeOrThrow(schema, operationType, field);
  const result = buildOperationTree(rootType, options, field);
  const selectionSet = parse(result, { noLocation: true }).definitions.map(definition => {
    if (definition.kind === 'OperationDefinition') {
      return definition.selectionSet;
    }

    return undefined;
  })[0];

  return {
    kind: 'Field',
    name: { kind: 'Name', value: field },
    alias: alias != null ? { kind: 'Name', value: alias } : undefined,
    arguments: operationArgs,
    selectionSet,
  };
};

/**
 * Builds a GraphQL Operation from a GraphQL Type.
 * @param schema The GraphQL Schema.
 * @param operationType The Operation Type.
 * @param operations The Operation(s) to build.
 * @returns An AST Document Node.
 */
export const buildOperation = (
  schema: GraphQLSchema,
  operationType: OperationTypeNode,
  ...operations: readonly BuildOperation[]
): DocumentNode => {
  const selections = operations.map(operation =>
    buildOperationSelectionNode(schema, operationType, operation)
  );

  const operationSelectionSet: SelectionSetNode = {
    kind: 'SelectionSet',
    selections,
  };
  const operationDefinition: OperationDefinitionNode = {
    kind: 'OperationDefinition',
    operation: operationType,
    selectionSet: operationSelectionSet,
  };

  return {
    kind: 'Document',
    definitions: [operationDefinition],
  };
};
