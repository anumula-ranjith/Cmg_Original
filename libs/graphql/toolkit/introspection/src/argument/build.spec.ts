import { assertIsNonNullable } from '@cmg/core-assert';
import type { ASTNode, IntrospectionQuery } from 'graphql';
import { assertInputType, buildClientSchema, GraphQLList, GraphQLNonNull, print } from 'graphql';

import introspectionJson from '../__mocks__/introspection.json';
import { buildArgumentNode, buildArgumentNodesByField } from './build';

const introspectSchema = introspectionJson.data as unknown as IntrospectionQuery;

describe('buildArgumentNodesByField', () => {
  it('builds a Query type argument nodes from a field', () => {
    const schema = buildClientSchema(introspectSchema);
    const typeName = assertIsNonNullable(schema.getQueryType()?.name);
    expect(typeName).not.toBeFalsy();
    const result = buildArgumentNodesByField({
      schema,
      typeName,
      field: 'offerings',
      args: {
        where: { accelerated: { eq: true } },
        order: { accelerated: 'ASC', company: { issuer: 'ASC' } },
      },
    });

    expect(print(result as unknown as ASTNode)).toMatchSnapshot();
  });
});

describe('buildArgumentNode', () => {
  it('builds an argument', () => {
    const schema = buildClientSchema(introspectSchema);
    const result = assertIsNonNullable(
      buildArgumentNode({
        schema,
        name: 'where',
        type: 'OfferingFilterInput',
        value: { accelerated: { eq: true } },
      })
    );

    expect(result).not.toBeUndefined();
    expect(print(result)).toMatchSnapshot();
  });

  it('builds an wrapped argument', () => {
    const schema = buildClientSchema(introspectSchema);
    const typeName = 'OfferingSortInput';
    const baseType = schema.getType(typeName);
    const scalarType = assertInputType(baseType);
    const type = GraphQLList(GraphQLNonNull(scalarType));
    const result = assertIsNonNullable(
      buildArgumentNode({
        schema,
        name: 'order',
        type,
        value: [{ accelerated: 'ASC' }, { company: { issuer: 'ASC' } }],
      })
    );

    expect(result).not.toBeUndefined();
    expect(print(result)).toMatchSnapshot();
  });
});
