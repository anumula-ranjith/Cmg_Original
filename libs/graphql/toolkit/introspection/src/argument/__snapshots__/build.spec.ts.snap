// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildArgumentNode builds an argument 1`] = `"where: {accelerated: {eq: true}}"`;

exports[`buildArgumentNode builds an wrapped argument 1`] = `"order: [{accelerated: ASC}, {company: {issuer: ASC}}]"`;

exports[`buildArgumentNodesByField builds a Query type argument nodes from a field 1`] = `
Array [
  "order: {accelerated: ASC, company: {issuer: ASC}}",
  "where: {accelerated: {eq: true}}",
]
`;
